### PR TITLE
Fix `SiPixelRawToClusterGPUKernel` for spurious ROCs [12.4.x]

### PR DESCRIPTION
**Pull request by fwyzard**
#### PR description:

Make the implementation of the pixel unpacker running on GPU skip spurious ROCs similar to the legacy version of the unpacker.

Fixes an HLT crash that has been happening more or less regularly for the past two months.

See #39045 for more details on the problem, investigation, and tests.

Also, changes the `debug` function parameter to a template parameter, to enable or disable the printouts at compile time.

#### PR validation:

Running the HLT over the error stream files does not crash any more.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Backport of #39711 to 12.4.x for data taking.

### CODE REVIEW

**Comment by fwyzard** [comment]
type bugfix

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
please test

File: RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.cu [hunk 1]
```diff
@@ -353,20 +351,25 @@ namespace pixelgpudetails {
...
         continue;
       }
 
-      uint32_t rawId = detId.rawId;
-      uint32_t rocIdInDetUnit = detId.rocInDet;
-      bool barrel = isBarrel(rawId);
+      // check for spurious channels
+      if (roc > MAX_ROC or link > MAX_LINK) {
+        if constexpr (debug) {
+          printf("spurious roc %d found on link %d, detector %d (index %d)\n", roc, link, rawId, gIndex);
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.cu:369]: rawId is "of course" wrong at this point...
one could use
```
getRawId(cablingMap, fedId, link, 1).rawId;
```
under the assumption that all roc in a link belong to the same det

Comment (fwyzard) [reply RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.cu]: Thanks.

**Comment by fwyzard** [comment]
please test

**Comment by VinInn** [comment]
```
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/748197d0f76ba5ceb2ca78d835294f77/opt/cmssw/el8_amd64_gcc10/cms/cmssw/CMSSW_12_4_X_2022-10-09-0000/src/HeterogeneousCore/CUDAServices/src/CUDAService.cc, line 151:
nvmlCheck(nvmlInitWithFlags(NVML_INIT_FLAG_NO_GPUS | NVML_INIT_FLAG_NO_ATTACH));
NVML Error 18: Driver/library version mismatch
```
????

**Comment by missirol** [comment]
please test

```
nvmlCheck(nvmlInitWithFlags(NVML_INIT_FLAG_NO_GPUS | NVML_INIT_FLAG_NO_ATTACH));
NVML Error 18: Driver/library version mismatch
```
Seen in previous days (https://github.com/cms-sw/cmssw/pull/39677#issuecomment-1272247033 , https://github.com/cms-sw/cmssw/pull/38761#issuecomment-1272896796), it appears to be a transient issue with the GPU PR tests.

FYI: @smuzaffar

**Comment by fwyzard** [comment]
> NVML Error 18: Driver/library version mismatch

means that (some of) the test machines need to reload the NVIDIA drivers.
@smuzaffar is there a way to know on which machine the tests were run ? or ask to reload the drivers on all of them ?

**Comment by smuzaffar** [comment]
@fwyzard , we do not have any dedicated GPU resource so we use HTCondor GPU resources for PR/IB tests. I will ask CERN IT to reload the drivers for these resources.

[a]
```
Building remotely on [grid121390](https://cmssdt.cern.ch/jenkins/computer/grid121390) 
gridid-121390.0 is-ssse3 disk-189G-164G is-avx512vl eos schedulername-bigbird20.cern.ch slc7_amd64 is-sse2 cpu-medium x86_64 is-avx singularity scripts user-cmsbuild nvidia cpu-tiny is-avx512bw kernel-3.10.0-1160.76.1.el7.x86_64 is-avx2 cpu-small cpu-4 is-avx512dq nvidia-515.43.04 is-avx512_vnni is-avx512f memory-15507 real-memory-15507 GenuineIntel is-sse4_2 auto-delete is-sse is-sse4_1 hostname-b7g47n9785 real-cpu-4 slc7 condor
```

**Comment by fwyzard** [comment]
OK, thanks.

**Comment by missirol** [comment]
urgent

TSG asked for a patch release including this PR, for HLT online operations.

@cms-sw/reconstruction-l2 , please prioritise the review of this PR (the `12_6_X` PR needs to be merged first, etc.).

**Comment by missirol** [comment]
please test

It's clear that this PR passes the tests. We can anyway try to 'clean up the labels' overnight. I admit I don't know the cause behind these recurring "RelVals-INPUT" errors; I wonder whether it is not solvable, or if there is an issue about it.

**Comment by rappoccio** [comment]
Status described here: https://github.com/cms-sw/cmssw/issues/39045#issuecomment-1279154838

**Comment by mandrenguyen** [comment]
+1

**Comment by perrotta** [comment]
+1

**Pull request merged**